### ISSUE TITLE
Extra metadata in the PermissionDenied error

### ISSUE THREAD

**Issue by skeggse** [ronf/asyncssh#453]
Coming from `paramiko`, I was hoping to inspect the `PermissionDenied` error when passing `preferred_methods=["none"]` to get the remote server's supported `auth_methods`, but I don't see an obvious way to do this without really reaching into `asyncssh`'s machinery. Any interest in supporting some extra metadata on the `PermissionDenied` error to include `auth_methods`, similar to how `paramiko` has an `allowed_types` field on `BadAuthenticationType`?

**Comment by ronf** [comment]
Rather than making a call where you need to pass `preferred_auth=['none']` and look for a list on the PermissionDenied exception, what do you think about instead having a `get_server_auth_methods()` call you could make, similar to `get_server_host_key()`? This would complete the key exchange and send the usual USERAUTH_REQUEST with the value of `'none'` like the code does now on a `connect()`, but stop as soon as it gets back that first USERAUTH_FAILURE with a list of allowed auth methods, closing the connection and then returning that list.

The use of `'none'` as an auth method is kind of an internal implementation detail and not documented as an allowed auth method. Also, this avoids the question of which auth methods to list when a real PermissionDenied happens, as there could have been multiple USERAUTH_FAILURE responses on the connection with different values for the allowed methods if the normal auth processing is allowed to proceed beyond the query for the initial list.

Would this address your use case?

**Comment by skeggse** [comment]
> Would this address your use case?

Yes, this would be great! Generally I'm happy to implement feature requests, though I'm not sure I have a strong sense for how to rearrange the private interfaces in a sensible manner. I did throw together a workaround that uses private interfaces, though 😅 

<details>
<summary>Example workaround script</summary>

```py
import asyncio
import asyncssh
from asyncssh.connection import SSHClientConnection, SSHClientConnectionOptions

class AuthError(Exception):
    def __init__(self, methods):
        super().__init__("authentication failed")
        self.methods = methods

def _process_userauth_failure(self, _pkttype: int, pktid: int,
                              packet: asyncssh.packet.SSHPacket) -> None:
    auth_methods = packet.get_namelist()
    partial_success = packet.get_boolean()
    packet.check_end()
    self._force_close(AuthError(auth_methods))

class CaptureSSHClientConnection(SSHClientConnection):
    _packet_handlers = {
        **SSHClientConnection._packet_handlers,
        # Override the packet handler for just the failure message.
        asyncssh.constants.MSG_USERAUTH_FAILURE:
        _process_userauth_failure,
    }

async def get_methods(hostname):
    loop = asyncio.get_event_loop()
    options = SSHClientConnectionOptions(
        username="root", preferred_auth=["none"])
    options.waiter = loop.create_future()
    try:
        await loop.create_connection(lambda: CaptureSSHClientConnection(
            loop, options=options, wait="auth"), hostname, 22)
        await options.waiter
    except asyncssh.PermissionDenied:
        raise Exception(
            "implementation regression: check that asyncssh's internal contract has not changed"
        )
    except AuthError as err:
        return err.methods
    return [b"none"]

print(asyncio.run(get_methods("host.example.com")))
```
</details>

**Comment by ronf** [comment]
Ok - I've got a first cut at this in commit 30b7fc1 in the "develop" branch if you'd like to give it a try. The new get_server_auth_methods() async call takes the same arguments as get_server_host_key() plus a username, since auth methods can vary depending on the user. It returns a sequence of strings of the available auth methods, automatically cleaning up the connection used to retrieve this data.

**Comment by skeggse** [comment]
Wow fantastic, thanks! This looks like it solves my exact use-case :)

I'll play around with it soon - haven't had a chance yet.

**Comment by ronf** [comment]
Glad to hear it - thanks for letting me know! This will be included in the next release...

**Comment by skeggse** [comment]
Sorry to bug you about this - do you have any sense for a timeline on the next release?

**Comment by ronf** [comment]
Sorry for the delay. Some other feature work ended up taking longer than I had planned, but I finally got that committed and have only a few small cleanup items remaining before the develop branch should be in good shape for a release. If all goes well, I should have that out later today.

**Comment by ronf** [comment]
This change is now available in AsyncSSH 2.10.0.
